### PR TITLE
[docs] Update style guidelines

### PR DESCRIPTION
**Pull request by yol**
## Description
Update code style guidelines: Make it more consistent (description and examples), group some similar things together, and add some more rules/hints.

## Motivation and Context
During reviews often things come up that we do not have any position on.

## Other
We may also want to have a short guide for Git commit messages; should that be part of this document or rather be separate?

I just tried to type something up to have a basis for discussion. Constructive criticism and feedback welcome :-)

### CODE REVIEW

**Comment by pkerling** [comment]
I pinged some people I know I've talked to about this or might be interested, but the discussion is not meant to be limited to those in any way.

**Review by Paxxi — APPROVED** [review]
Some C++14 is used in windows only code as msvc haven't adhered to strict versioning flags. The one that comes to mind is `make_unique`. Not sure if it's worth mentioning as it might be confusing.

Nice work 😀👍

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: Can you add a newline between the ifs?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: "Always include a `default` case, unless switching on an enum and all enum values are handled."

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: In many parts of the code, `case` isn't indented to reduce the number of nested indentations

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: I know we've discussed left-aligning or right-aligning, what do other projects/people do and what is their argument?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: This brings up two points - we want to avoid const on primitive parameters, e.g. `test(const int a)`, and also we allow const after the modifier when we actually want const values, e.g.

```cpp
class Foo
{
  const Bar* const m_immutable;
};
```

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: Most of our code doesn't include this newline

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: We generally put other libraries' headers before system files

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: We want to move to `constexpr`

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: We want to use `enum class`

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: You should mention that overriding methods shouldn't use `virtual`, only `override`.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: We don't use this, but I like it.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: No new globals, period.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: Maybe include an example where the parameters and return type are doxy'd?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: We should discourage `auto` when it isn't necessary for readability, as it is opaque compared to using the actual types.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: Second sentence needs a comma

**Comment by garbear** [comment]
Nice. I left all the comments I had.

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: > This brings up two points - we want to avoid const on primitive parameters, e.g. test(const int a)

Don't have a strong opinion on this one - I think it's fine to do so if you want to emphasize that the function should not change the variable in its body (even if it would not be visible to the caller anyway).

> and also we allow const after the modifier when we actually want const values

Yeah that's a different type of `const` but maybe I can improve the wording.

File: docs/CODE_GUIDELINES.md
Comment (Paxxi) [reply docs/CODE_GUIDELINES.md]: I prefer using auto whenever the type is obvious from context

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: We shouldn't allow that even if right-aligning. It's confusing either way.

Anyway, https://github.com/xbmc/xbmc/pull/14665/files#diff-6c2710b7aab11c31227a420f277f49c7R242

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: Yeah I even forgot about that in one example :-D

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: Oh snap, you're right. Must have mixed it up with some other project. Anyway I can leave it in if people think it helps.

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: What about when they're local to a file?

e.g. https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/wayland/WinEventsWayland.cpp#L208

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: > Maybe include an example where the parameters and return type are doxy'd?

Good idea. Do you have one in mind?

> I saw below, I think that parameters should be explicitly documented, even if it's trivial (though trivial parameter descriptions are discouraged, of course)

Not completely sure, but I tend to disagree. Comments should only be there if they add value. What value do you think is added by having a doxy comment like in the (intentionally bad) example?

Comment (Paxxi) [reply docs/CODE_GUIDELINES.md]: Messaging/helpers should have parameters and return types documented.
Appmessenger and surrounding code is well documented (quality is debatable 😀)

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: ¯\_(ツ)_/¯

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: With @garbear on this one

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-auto
> Exception: Avoid auto for initializer lists and in cases where you know exactly which type you want and where an initializer might require conversion.

https://google.github.io/styleguide/cppguide.html#auto
> Use auto to avoid type names that are noisy, obvious, or unimportant - cases where the type doesn't aid in clarity for the reader. Continue to use manifest type declarations when it helps readability.

But maybe that's actually also what you mean with "obivous from context" @Paxxi ?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: I see you included that requirement. Thanks for looking up the existing guidelines, maybe sight these examples and list our reasoning for other projects that are having the same debate?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: > Don't have a strong opinion on this one - I think it's fine to do so if you want to emphasize that the function should not change the variable in its body (even if it would not be visible to the caller anyway).

I am against this, because I don't think this internal detail should be exposed to external callers. However, inside a function, I encourage the use of const variables as much as possible -- see how rust defaults everything to const and requires `mutable` to be declared explicitly.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: That's not *too* offensive, because `std::unique_ptr` construction and destruction is trivial. However, I'm greatly concerned that `g_WlMessagePump` isn't reset() deterministically, because thread destructors are extremely non-trivial.

I stand by my no-global requirement, because they can always be encapsulated as a static variable in a class with no downside.

Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: > I stand by my no-global requirement, because they can always be encapsulated as a static variable in a class with no downside.

What's the upside?

File: docs/CODE_GUIDELINES.md
Comment (Paxxi) [reply docs/CODE_GUIDELINES.md]: Yeah, if a method is named `GetFileItems` I don't need to spell out that it's a `vector<CFileItem>`

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: In this case, my argument isn't for the result but the habit - developers should practice documenting everything. Oftentimes, I'll document three parameters, the first two will be trivial but the third I expand upon. The "noise" from the first two is worth the reminder to myself to expand upon the more opaque parameter.

It's like including the many headers in our PR templates - the dev is free to ignore and delete any headers, but I want them to be conscious that they're doing so. It should take extra effort to document less, not the other way around.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: What's obvious to you may not be obvious to a new dev unfamiliar with the code.

**Comment by pkerling** [comment]
> Some C++14 is used in windows only code as msvc haven't adhered to strict versioning flags. The one that comes to mind is make_unique. Not sure if it's worth mentioning as it might be confusing.

I wouldn't mention it at the moment, it would just confuse people. I kind of hope that we'll be switching to C++14 for v19 anyway :pray:

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: > Messaging/helpers should have parameters and return types documented.

Tried a utils one.

Related question: Do we want to have the `*` on the left side of the doxy comments? I kind of like them because it makes it more visually distinct in the editor, but I'm open for opinions.

> In this case, my argument isn't for the result but the habit - developers should practice documenting everything.

I'm seeing the opposite problem - if you get into the habit of "documenting" everything by adding boilerplate text that does not add information, you not only clutter the code, but may also grow into an automatism that prevents you from writing actually meaningful comments.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: Here's another point, when the global is owned by a class, it's very clear who is in charge of creating/destroying the global. In your cited example, it is less clear that `CWinEventsWayland` is responsible for creating and destroying the object (and indeed, fails to destroy it deterministically).

**Comment by pkerling** [comment]
Would it be a good idea to advocate CLog::Log("{}", 123) as opposed to CLog::Log("%d", 123) for type safety reasons?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: Yeah, because the HP guidelines explicitly state the "what" and "why", I would insert similar to what you said above:

> HP C++ Coding Guidelines say left-align with this reasoning: "The characters * and & are to be written with the type of variables instead of with the name of variables in order to emphasize that they are part of the type definition."

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: What are others' opinions?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: It isn't about knowing that size_t is 64bit, it's about knowing that vec.size() returns a size_t. Obviously most devs know this, but being explicit is explanatory to new devs.

**Comment by garbear** [comment]
> Would it be a good idea to advocate CLog::Log("{}", 123) as opposed to CLog::Log("%d", 123) for type safety reasons?

Most definitely. Is our new log format method documented anywhere?

**Comment by Paxxi** [comment]
It's documented here http://fmtlib.net/latest/syntax.html

File: docs/CODE_GUIDELINES.md
Comment (Paxxi) [reply docs/CODE_GUIDELINES.md]: That is true. I don't believe it to be an issue in practice though. And when in doubt your ide will display the type on hover or similar method.
I'm not sure it makes sense to force a style optimized for simple text editor users.
Either don't mention auto or say something like, use your judgment and when in doubt write the type

**Comment by notspiff** [comment]
Fuck ide's, why would anyone use those things?

Den tir. 23. okt. 2018, 08:03 skrev Pär Björklund <notifications@github.com

**Comment by Paxxi** [comment]
Does cim/emacs surface the info in an easy manner with the right plug-ins?

**Comment by notspiff** [comment]
Irrelevant. My reaction is due to the ide statement. Some use them, some
dont, don't use the code guide lines to enforce either.

Den tir. 23. okt. 2018, 08:37 skrev Pär Björklund <notifications@github.com

**Comment by Paxxi** [comment]
Okay. You appear to have strong feelings on this so I accept that. Go ahead with @garbear original comments.

File: docs/CODE_GUIDELINES.md
Comment (DaveTBlake) [reply docs/CODE_GUIDELINES.md]: Also with garbear on readability (I read with my eyes, not pointing the mouse)

**Comment by notspiff** [comment]
I don't care much about auto usage at but arguments based on 'everybody
does as me' do trigger me.

Den tir. 23. okt. 2018, 08:50 skrev Pär Björklund <notifications@github.com

File: docs/CODE_GUIDELINES.md
Comment (DaveTBlake) [reply docs/CODE_GUIDELINES.md]: Can we clarify if sort order is case sensitive or not. The example implies a...z,A...Z, but good to state it explicitly if that is what we want. I don't mind which and I have seen both in existing code (or maybe it was just a jumble).

File: docs/CODE_GUIDELINES.md
Comment (DaveTBlake) [reply docs/CODE_GUIDELINES.md]: Having a newline would make it  a consistent separator of sorted grosups, but not a strong opinion

**Comment by lrusak** [comment]
@pkerling I don't have any strong feeling either way. I'd rather let the experts determine this stuff and I'll follow with whatever is decided (as long as it's for good reason) 👍 

Also, here is the rich text for those who want to see what it actually looks like 💯 
https://github.com/pkerling/xbmc/blob/code-style-guidelines/docs/CODE_GUIDELINES.md

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: Instead of:

> However, do not describe trivialities - it only adds visual noise.

How about:

> Avoid trivialities - this is a sign that you should be more descriptive.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: Can you add a newline between this and the params? Also sometimes I add a newline between params and return, but our style guide doesn't necessarily need this.

File: docs/CODE_GUIDELINES.md
Comment (garbear) [inline docs/CODE_GUIDELINES.md]: lol

File: docs/CODE_GUIDELINES.md
Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: > Can we clarify if sort order is case sensitive or not. The example implies a...z,A...Z,

I don't think this is implied by the example. It explicitly says: "Place directories before files", which is I think the rule that is followed. Do we actually have cases where case matters (pun not intended) after putting dirs first?

File: docs/CODE_GUIDELINES.md
Comment (garbear) [reply docs/CODE_GUIDELINES.md]: Sure, your wording is fine.

Comment (pkerling) [reply docs/CODE_GUIDELINES.md]: I get what you want to say though. If you find a wording that combines both, let me know...

File: docs/CODE_GUIDELINES.md
Comment (DaveTBlake) [reply docs/CODE_GUIDELINES.md]: Doh, I have clearly never understood what "Place directories before files" meant! But that does make sense of what I have seen, sorry for noise

**Comment by pkerling** [comment]
#14672 has seen some discussion on initialization with `{}` that I'd like to respond to and I feel would be better suited for this PR.

>  And I don't like the init of primitive types with () or {}. Especially if no truncation can occur which wasn't intended.

Initialization with `()` is off the table anyway. The other is for consistency. You shouldn't have to think about whether you could truncate something or not, you should just get an error if you do. And you should explicitly `static_cast` for truncating so that it is clear it is what you want. That's my opinion at least.

> Value/zero-initialization via braces for primitive types looks weird IMO.

As @garbear already mentioned, this is purely a matter of habit and I don't think we should decide based on that. I did try not to push this too hard in the style guidelines though and only *recommend* using `{}`. C++ core guidelines also acknowledge this: "Old habits die hard, so this rule is hard to apply consistently, especially as there are so many cases where = is innocent."

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-list

> When skimming through code like that, you have to also look at the type (worse when it's a typedef to a primitive) and you have to know that primitive types are value-initialized to 0.

The latter part does not seem very relevant, it's also mostly a matter of being in the habit. You'll know about it after you've seen it 2, 3 times.

Do note that I am *not* arguing against writing `int a{0}` (or similar) at all. It's all about intent. If you feel it makes the code more readable when the `0` is there - by all means do put it there. If it's clear anyway and would be superfluous - leave it out. Maybe this is not clear from the description in the style guide?

> Granted, every sane person would assume that, but it's extra mental effort for no actual reason besides using more modern syntax and 2 less characters. C++ has a bajillion ways to initialize shit, don't make it more complicated

Reiterating: `{[...]}` is applicable pretty much anywhere, so it is the one initialization solution that has the potential to make it *less* complicated - if applied consistently.

I do know that this will be hard to impossible to achieve completely with the Kodi code base (like most of these rules really). So I'd in principle be fine with saying we recommend the `int a = 1` syntax (and ignore narrowing and other problems) for primitive types because `{}` not very popular yet even though it's technically superior.

**Comment by DaveTBlake** [comment]
>So I'd in principle be fine with saying we recommend the `int a = 1` syntax (and ignore narrowing and other problems) for primitive types because {} not very popular yet even though it's technically superior.

I would prefer that - after nearly 40 year of reading code in many different languages, a simple `=` is so easy to read and understand in variable initialisation.

**Comment by garbear** [comment]
I second @DaveTBlake - a simple `=` is easier to read and understand. And that's what's most important. Maybe in a few years `{}` will take over, but now I think our rule should be `{}` is for non-primitive types, and `=` for primitive types is allowed when it improves readability.

**Comment by Paxxi** [comment]
I don't think we should be too strict here. Goal IMO should be to get contributors to do the right thing and not face nitpicking over style when submitting a pr.
Allowing common styles seems fine to me unless they cause serious readability issues.
If say we want to force init with {} in the future we should run clang-tidy on everything so new contributors only see the approved style and get it right by following surrounding code style.

**Comment by pkerling** [comment]
OK, so the direction seems to be to prefer `=` for zero and explicit initialization of primitive types, but prefer `{}` for non-primitive types (so you'd write e.g. `CPointInt p1{3, 5}` and not `CPointInt p1(3, 5)`). Did I get that right? I'm not sure the last point would be very helpful, since the upside of `{}` is mostly that you can use it for everything - so limiting it to one specific case kind of defeats its purpose. It still gets you around most vexing parse, though.

And do we want to make that a fixed rule for new code or allow both alternatives by only giving a recommendation?

**Comment by Paxxi** [comment]
I would make it a recommendation. Both have upsides and drawbacks so allowing both is most pragmatic.

**Comment by MartijnKaijser** [comment]
Can we merge whatever we have now?

**Comment by pkerling** [comment]
@MartijnKaijser I'd go over it once more and merge for v19. That ok?

Haven't forgotten about it, just busy atm and this isn't high priority.

**Comment by pkerling** [comment]
> I would make it a recommendation.

Done - ready to merge from my side. Everybody happy? :-)

**Comment by MartijnKaijser** [comment]
squash? :)

**Comment by pkerling** [comment]
After I got some +1s and am sure I don't have to edit something again, yes ;-D

**Review by a1rwulf — APPROVED** [review]
I'm fine with it.

**Comment by pkerling** [comment]
Thanks all for the fruitful discussion!

**Pull request merged**
